src/status/libstatus/coder.nim

```diff
@@ -0,0 +1,185 @@
+import sequtils, strformat, sugar, macros, tables, strutils
+import eth/common/eth_types, stew/byteutils
+
+type
+  FixedBytes* [N: static[int]] = distinct array[N, byte]
+  DynamicBytes* [N: static[int]] = distinct array[N, byte]
+  EncodeResult* = tuple[dynamic: bool, data: string]
+  Bool* = distinct Int256 # TODO: implement Bool as Stint[256]?
+  Encodable* = concept x
+    encode(x) is EncodeResult
+
+type
+  GetPackData* = object
+    packId*: Stuint[256]
+
+  PackData* = object
+    category*: DynamicBytes[32] # bytes4[]
+    owner*: EthAddress # address
+    mintable*: bool # bool
+    timestamp*: Stuint[256] # uint256
+    price*: Stuint[256] # uint256
+    contentHash*: DynamicBytes[64] # bytes
+
+  BuyToken* = object
+    packId*: Stuint[256]
+    address*: EthAddress
+    price*: Stuint[256]
+
+  ApproveAndCall* = object
+    to*: EthAddress
+    value*: Stuint[256]
+    data*: DynamicBytes[100]
+
+  Transfer* = object
+    to*: EthAddress
+    value*: Stuint[256]
+
+  BalanceOf* = object
+    address*: EthAddress
+
+  TokenOfOwnerByIndex* = object
+    address*: EthAddress
+    index*: Stuint[256]
+
+  TokenPackId* = object
+    tokenId*: Stuint[256]
+
+  TokenUri* = object
+    tokenId*: Stuint[256]
+
+proc skip0xPrefix*(s: string): int =
+  if s.len > 1 and s[0] == '0' and s[1] in {'x', 'X'}: 2
+  else: 0
+
+proc strip0xPrefix*(s: string): string =
+  let prefixLen = skip0xPrefix(s)
+  if prefixLen != 0:
+    s[prefixLen .. ^1]
+  else:
+    s
+
+proc fromHexAux*(s: string, result: var openarray[byte]) =
+  let prefixLen = skip0xPrefix(s)
+  let meaningfulLen = s.len - prefixLen
+  let requiredChars = result.len * 2
+  if meaningfulLen > requiredChars:
+    let start = s.len - requiredChars
+    hexToByteArray(s[start .. s.len - 1], result)
+  elif meaningfulLen == requiredChars:
+    hexToByteArray(s, result)
+  else:
+    raise newException(ValueError, "Short hex string (" & $meaningfulLen & ") for Bytes[" & $result.len & "]")
+
+func fromHex*[N](x: type FixedBytes[N], s: string): FixedBytes[N] {.inline.} =
```

Comment:
How does this code differ from the existing one in nim-web3? Why did you decide to create duplicate definitions here instead of reusing or extending the ones provided there?